sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -220,6 +222,404 @@ case class EndsWith(left: Expression, right: Expression)
+        val p = pad.eval(input)
+        if (p == null) {
+          null
+        } else {
+          val len = l.asInstanceOf[Int]
+          val str = s.asInstanceOf[UTF8String]
+          val pad = p.asInstanceOf[UTF8String]
+          val bytes = new Array[Byte](len)
+
+          performOp(bytes, str.getBytes, pad.getBytes, len, str, pad)
+        }
+      }
+    }
+  }
+
+  // Copied from org.apache.hadoop.hive.ql.udf.generic.GenericUDFRpad
+  // ('hi', 5, '??') => 'hi???'
+  // ('hi', 1, '??') => 'h'
+  protected def performOp(
+    data: Array[Byte],
+    txt: Array[Byte],
+    padTxt: Array[Byte],
+    len: Int,
+    str: UTF8String,
+    pad: UTF8String): UTF8String = {
+
+    // Copy the text
+    var pos = 0
+    while (pos < str.length() && pos < len) {
+      data(pos) = txt(pos)
+      pos += 1
+    }
+
+    // Copy the padding
+    while (pos < len) {
+      var i = 0
+      while (i < pad.length() && i < len - pos) {
+        data(pos + i) = padTxt(i)
+        i += 1
+      }
+
+      pos += pad.length()
+    }
+
+    UTF8String.fromBytes(data)
+  }
+
+  override def toString: String = s"RPAD($str, $len, $pad)"
+}
+
+/**
+ * Returns the input formatted according do printf-style format strings
+ */
+case class StringFormat(children: Expression*) extends Expression {
+
+  require(children.length >=1, "printf() should take at least 1 argument")
+
+  override def foldable: Boolean = children.forall(_.foldable)
+  override def nullable: Boolean = children(0).nullable
+  override def dataType: DataType = StringType
+  private def format: Expression = children(0)
+  private def args: Seq[Expression] = children.tail
+
+  override def eval(input: InternalRow): Any = {
+    val pattern = format.eval(input)
+    if (pattern == null) {
+      null
+    } else {
+      val sb = new StringBuffer()
+      val formatter = new java.util.Formatter(sb, Locale.US)
+
+      val arglist = args.map(_.eval(input).asInstanceOf[AnyRef])
+      formatter.format(pattern.asInstanceOf[UTF8String].toString(), arglist: _*)
+
+      UTF8String.fromString(sb.toString)
+    }
+  }
+
+  override def toString: String = s"printf($format, $args)"
+}
+
+/**
+ * Returns the string which repeat the given string value n times.
+ */
+case class StringRepeat(str: Expression, times: Expression)
+  extends Expression with ExpectsInputTypes {
+
+  override def children: Seq[Expression] = str :: times :: Nil
+  override def foldable: Boolean = str.foldable && times.foldable
+  override def nullable: Boolean = str.nullable || times.nullable
+  override def dataType: DataType = StringType
+  override def expectedChildTypes: Seq[DataType] = Seq(StringType, IntegerType)
+
+  override def eval(input: InternalRow): Any = {
+    val s = str.eval(input)
+    if (s == null) {
+      null
+    } else {
+      val t = times.eval(input)
+      if (t == null) {
+        null
+      } else {
+        UTF8String.fromString(s.asInstanceOf[UTF8String].toString() * t.asInstanceOf[Integer])
```

Comment:
Can we just repeat `UTF8String.bytes` n times and create a new `UTF8String` using new bytes?